pallets/providers/src/types.rs

```diff
@@ -0,0 +1,121 @@
+    traits::fungible,
+};
+use scale_info::prelude::fmt::Debug;
+use scale_info::TypeInfo;
+
+use crate::Config;
+
+/// Structure that has the possible value propositions that a main storage provider can offer (and the runtime is aware of)
+#[derive(Encode, Decode, MaxEncodedLen, TypeInfo, RuntimeDebugNoBound, PartialEq, Eq, Clone)]
+#[scale_info(skip_type_params(T))]
+pub struct ValueProposition<T: Config> {
+    pub data_limit: StorageData<T>,
+    pub protocols: BoundedVec<Protocols<T>, MaxProtocols<T>>,
+    // TODO: Add relevant fields here
+}
+
+/// Structure that represents a main storage provider. It holds the amount of data that the MSP is able to store,
+/// the amount of data that it IS storing, its libp2p multiaddress, and its value proposition.
+#[derive(Encode, Decode, MaxEncodedLen, TypeInfo, RuntimeDebugNoBound, PartialEq, Eq, Clone)]
+#[scale_info(skip_type_params(T))]
+pub struct MainStorageProvider<T: Config> {
+    pub total_data: StorageData<T>,
+    pub data_used: StorageData<T>,
+    pub multiaddress: MultiAddress<T>,
+    pub value_prop: ValueProposition<T>,
+}
+
+/// Structure that represents a backup storage provider. It holds the amount of data that the BSP is able to store,
+/// the amount of data that it is storing, and its libp2p multiaddress.
+#[derive(Encode, Decode, MaxEncodedLen, TypeInfo, RuntimeDebugNoBound, PartialEq, Eq, Clone)]
+#[scale_info(skip_type_params(T))]
+pub struct BackupStorageProvider<T: Config> {
+    pub total_data: StorageData<T>,
+    pub data_used: StorageData<T>,
+    pub multiaddress: MultiAddress<T>,
+}
+
+// Type aliases:
+
+/// BalanceOf is the balance type of the runtime.
+pub type BalanceOf<T> =
+    <<T as Config>::NativeBalance as Inspect<<T as frame_system::Config>::AccountId>>::Balance;
+
+/// MultiAddress is a byte array that represents the libp2p multiaddress of a storage provider.
+/// Its maximum size is defined in the runtime configuration, as MaxMultiAddressSize.
+pub type MaxMultiAddressSize<T> = <T as crate::Config>::MaxMultiAddressSize;
+pub type MultiAddress<T> = BoundedVec<u8, MaxMultiAddressSize<T>>;
+
+/// StorageData is the type of the unit in which we measure data size. We define its required traits in the
+/// pallet configuration so the runtime can use any type that implements them.
+pub type StorageData<T> = <T as crate::Config>::StorageData;
+
+/// Protocols is a vector of the protocols that (the runtime is aware of and) the main storage provider supports.
+/// Its maximum size is defined in the runtime configuration, as MaxProtocols.
+pub type MaxProtocols<T> = <T as crate::Config>::MaxProtocols;
+pub type Protocols<T> = BoundedVec<u8, MaxProtocols<T>>; // TODO: define a type for protocols
+
+/// MaxBsps is the maximum amount of backup storage providers that can exist. It is defined in the runtime configuration.
+pub type MaxBsps<T> = <T as crate::Config>::MaxBsps;
+
+/// Trait definitions:
```

Comment:
"Trait definitions" is going to show as part of `StorageProvidersInterface` docs here